`neil new`: Add support for deps-based templates

rads opened the pull request:
## Usage

Let's say you want to set up a `kit-clj` project on a new macOS machine. Here's what that looks like without `neil`:
```shell
brew install clojure
clojure -Ttools install com.github.seancorfield/clj-new '{:git/tag "v1.2.381"}' :as new
clojure -Tnew create :template io.github.kit-clj :name yourname/app
```

And then with `neil`:
```shell
brew install clojure babashka/brew/neil
neil new io.github.kit-clj/kit yourname/app
```

## Testing

**Note:** To make the above command work temporarily, you need to add the `--git/url` and `--git/sha` options to use my [fork of kit-clj/kit](https://github.com/rads/kit#quick-start):
```
./neil new io.github.kit-clj/kit yourname/app \
  --git/url https://github.com/rads/kit \
  --git/sha cb2236503c5fd5202df28bca1551bd2128a71e23
```
Once Kit gets support for `deps-new` (see https://github.com/kit-clj/kit/issues/52), the `--git/url` and `--git/sha` options can be removed.

## Notes

If you run into an error like this while testing, this is due to hitting the GitHub rate limit:
```
Error building classpath. Library io.github.rads/kit has coord with missing sha
```

## TODO

- [x] Add tests
  - [x] Unit test for `neil new scratch ...`
  - [x] End-to-end test for `neil new io.github.rads/neil-new-test-template ...`
    - [ ] Change `io.github.rads/neil-new-test-template` to something else?
  - [x] Unit tests for extended options
- [x] Update docs

rads commented on src/babashka/neil.clj:
```diff
@@ -381,11 +384,29 @@
+  (str "https://github.com/" (clean-github-lib lib)))
+
+(defn- add-template-deps [template opts]
+  (let [lib (edn/read-string template)
```
Since it's possible to have multiple `template.edn` files in the same `deps.edn` project, it might be worth having a separate override option for the `lib` instead of inferring the `lib` from the `template` all the time.

borkdude replied:
I'm not exactly sure what you mean by this. Could you elaborate or point to a test / doc updates in this PR which explain this?

rads replied:
No problem. See [this example from the deps-new README](https://github.com/seancorfield/deps-new#more-general-usage). I split it into multiple lines for readability:
```shell
clojure \
  -Sdeps '{:deps {com.acme.project/cool-lib COORDINATES}}' \
  -Tnew create \
  :template com.acme.project/cool-lib \
  :name myusername/mynewproject
```
In theory you could change the first instance of `com.acme.project/cool-lib` to a different name and the example would still work as long as the directory structure in `:paths` of the lib matches the `:template` option (`resources/com/acme/project/cool_lib`).

This could be useful if you had multiple templates in the same lib which you could then switch between using `:template`. That said, it's possible to achieve the same thing by making a separate lib for each template, so it may not be necessary to support this.

borkdude replied:
Let's not support things we're not certain we're going to need, so let's wait with this.

rads commented on src/babashka/neil.clj:
```diff
@@ -364,6 +362,11 @@
+   ; template deps overrides
+   :local/root
+   :git/url
+   :git/sha
```
From the [deps-new docs](https://github.com/seancorfield/deps-new#more-general-usage):
> `:local/root` and `git`-based coordinates are supported, but Maven/Clojars coordinates are not.

rads commented on tests.clj:
```diff
@@ -82,5 +82,23 @@
+    (spit (test-file "deps.edn") "{}")
+    (neil (str "new scratch my-scratch :overwrite true :target-dir " target-dir))
+    (is (= (slurp (fs/file "test-resources/new/my-scratch/src/scratch.clj"))
+           (slurp (fs/file (str target-dir "/src/scratch.clj")))))
```
Right now I'm using static files in the `test-resources` directory to make the tests work. Unfortunately this will cause the tests to break if the template in `deps-new` is changed.

One way to work around this would be to add a command-line option to swap out the `org.corfield.new/create` function with a mock during testing. That said, I wanted to get some feedback first before adding something like that.

rads replied:
Thinking about this some more... maybe a `--dry-run` feature would be better than a mock. Then we can read back the output to check that the `org.corfield.new/create` is being passed the parameters we expect.

rads replied:
Added `:dry-run` option.

rads commented:
Made some progress on porting Kit's existing `clj-new` template to `deps-new`: https://github.com/kit-clj/kit/issues/52#issuecomment-1193373127

borkdude commented on neil:
```diff
@@ -379,6 +377,12 @@
+   :dry-run
+
+   ; template deps overrides
+   :local/root
```
It's not necessary to explicitly list all the options. Babashka CLI automatically parses all command line flags. The list your using here is for converting options in order to options, using `:cmds-opts`.

borkdude replied:
This is just a note.

rads replied:
Makes sense. I updated the `:cmds-opts` to be `[:template :name :target-dir]`. These are the ones that made most sense to me as positional arguments.

For the special named arguments which are not directly supported by `deps-new` (`:git/url`, `:git/sha`, `:git/tag`, `:local/root`), we'll probably want to add those to the `neil help` docs in addition to the existing link to the `deps-new` override options.

borkdude replied:
On line 16 (on main) we already have a bunch of options. I'd prefer if we could align the names of options, even if they are used in different contexts. Since we already have `:sha` , maybe we should re-use that one? Or should we deprecate it and rename it to `:git/sha` (with `:sha` being an `:alias` for `:git/sha`)?

rads replied:
@borkdude: Here's the solution I came up with:

- `neil new` will always accept namespaced keys like `:git/sha`
- `neil new` now also supports the existing `:sha` and `:latest-sha` options
  - `:sha` is simply renamed to `:git/sha` before parsing
  - `:latest-sha` is a useful new option that provides an alternative to fetching the latest `:git/tag` (which is the default behavior when neither `:git/tag`, `:git/sha`, nor `:latest-sha` is provided)
- Any options for `neil new` that are not already in `babashka.neil/spec` should be documented specifically for `neil new`. If we find some `neil new` options to be useful elsewhere, we can promote them to `babashka.neil/spec` as needed.

What do you think about this?

borkdude commented on neil:
```diff
@@ -397,12 +401,46 @@
+     :create-opts create-opts}))
+
+(defn- deps-new-create [create-opts]
+  ((resolve 'org.corfield.new/create) create-opts))
```
We can move requiring-resolve call over here in this function

rads commented:
I updated the logic in `template-libs` so we can fail fast when detecting invalid combinations of deps options, such as `:local/root` and `:git/url` at the same time. I split the logic into separate cases since it was getting tricky to implement everything in one place.

rads commented:
Dev note: Not sure if there's a better way to do this, but here's what I'm currently using to auto-build the `neil` file during development (using [entr](http://eradman.com/entrproject/)):
```
find . -iname '*.clj' -o -iname 'prelude' | entr -s 'bb gen-script && date'
```

rads commented:
@borkdude: Finished up with everything I originally planned to do so far. This is officially ready for review. 🙂

rads commented:
I realized I hadn't tested this with `clj -M:neil` yet and I found some bugs with that. I just pushed up a commit to make the `new` command work in both `bb` and `clj` environments.

One caveat for `clj -M:neil` usage is that `babashka.deps/add-deps` is not supported, so custom templates need to be specified explicitly in a `deps.edn` file or `-Sdeps`. Seems like a reasonable tradeoff to me since that's how `deps-new` already works. I think most people are going to want to use the `bb` environment anyways when they don't want to specify `deps.edn`:

```
clojure -M:neil new io.github.rads/neil-new-test-template foo :overwrite   5.62s user 0.19s system 294% cpu 1.973 total
neil new io.github.rads/neil-new-test-template foo    0.10s user 0.07s system 30% cpu 0.523 total
```

rads commented on src/babashka/neil.clj:
```diff
@@ -357,36 +364,190 @@
+
+https://github.com/seancorfield/deps-new/blob/develop/doc/options.md
+
+Both built-in and remote templates are supported. Built-in templates use
```
Maybe "remote" is not the right word here since `:local/root` can be used?

borkdude replied:
Maybe `external` is a better word?

rads replied:
Yep, that sounds better to me!

borkdude commented on neil:
```diff
@@ -373,36 +380,195 @@
+  "Returns parsed deps-new template deps opts from raw CLI opts."
+  [cli-opts]
+  (-> cli-opts
+      (set/rename-keys {:sha :git/sha})
```
This we can do via the `:alias` feature of babashka.cli I think?
